### DISCUSSION TITLE
TamperMonkey Userscript - CVSS String

### DISCUSSION THREAD

**Discussion by leonteale** [pwndoc/pwndoc#493] [Show and tell]
Ok so something I wanted the app to do but it doesnt, is show the CVSS string on the page when you set the CVSS score.  So I made a tamper monkey to do it. I am not very good with github so dont know if it's something PwnDoc wants to have from me or not?  for anyone interested here is the code (or message me for the link as it's on my github): [https://github.com/leonteale/pentestpackage/tree/master/TamperMonkeyScripts](url)

hopefully this loads the image as an example. 
[![2023-08-24-18-19-49-pwndoc-and-6-more-pages-Work-Microsoft-Edge.png](https://i.postimg.cc/26Z3Y9RS/2023-08-24-18-19-49-pwndoc-and-6-more-pages-Work-Microsoft-Edge.png)](https://postimg.cc/47XXb5xD)

```
// ==UserScript==
// @name         PwnDoc CVSS String Builder
// @namespace    http://cyberwolf-security.co.uk/
// @version      1.0
// @description  Build CVSS string for PwnDoc
// @author       Leon Teale
// @match        https://10.200.200.5:8443/*
// @grant        none
// ==/UserScript==

(function() {
    'use strict';

    // Define the sections and their corresponding codes
    const sections = [
        { name: 'Attack Vector', code: 'AV' },
        { name: 'Attack Complexity', code: 'AC' },
        { name: 'Privileges Required', code: 'PR' },
        { name: 'User Interaction', code: 'UI' },
        { name: 'Scope', code: 'S' },
        { name: 'Confidentiality Impact', code: 'C' },
        { name: 'Integrity Impact', code: 'I' },
        { name: 'Availability Impact', code: 'A' },
    ];

    // Mapping between option text and CVSS code
    const mapping = {
        "Network": "N",
        "Adjacent Network": "A",
        "Local": "L",
        "Physical": "P",
        "High": "H",
        "Low": "L",
        "None": "N",
        "Required": "R",
        "Unchanged": "U",
        "Changed": "C",
    };

    // Function to get the CVSS value for a given section name
    function getCVSSValue(name) {
        const span = Array.from(document.querySelectorAll('span')).find(span => span.textContent.includes(name));
        if (span) {
            const button = span.parentElement.nextElementSibling.querySelector(`button[aria-pressed='true']`);
            return button ? button.textContent.trim() : null;
        }
        return null;
    }

    // Function to build the CVSS string based on selected values
    function buildCVSSString() {
        let cvssString = ''; // Keep the label
        for (const section of sections) {
            const value = getCVSSValue(section.name);
            const code = mapping[value] || 'X';
            cvssString += `${section.code}:${code}/`;
        }
        displayCVSSString(cvssString.slice(0, -1)); // Remove trailing slash
    }

    // Function to display the CVSS string on the page
    function displayCVSSString(cvssString) {
        const container = document.querySelector('div.col-md-3');
        if (!container) return; // Check if container exists
        let cvssElement = container.querySelector('.cvss-string');
        if (!cvssElement) {
            cvssElement = document.createElement('span');
            cvssElement.className = 'cvss-string';
            cvssElement.style.color = 'yellow';
            container.appendChild(cvssElement);
        }
        cvssElement.textContent = cvssString;
    }

    // Initialize the script by attaching click listeners and building the initial CVSS string
    function init() {
        const buttons = document.querySelectorAll('.q-btn-group button');
        buttons.forEach(button => button.addEventListener('click', () => {
            setTimeout(buildCVSSString, 100); // Delay to allow the button state to update
        }));
        buildCVSSString();
    }

    // Use a MutationObserver to reinitialize the script if the page content changes
    const observer = new MutationObserver(init);
    observer.observe(document, { childList: true, subtree: true });
    init();
})();
```

**Comment by Tushargupta12345** [comment]
// ==UserScript==
// @name         PwnDoc CVSS String Builder
// @namespace    http://cyberwolf-security.co.uk/
// @version      1.0
// @description  Build CVSS string for PwnDoc
// @author       Leon Teale
// @match        https://10.200.200.5:8443/*
// @grant        none
// ==/UserScript==

(function() {
    'use strict';

    // Define the sections and their corresponding CVSS codes
    const sections = [
        { name: 'Attack Vector', code: 'AV' },
        { name: 'Attack Complexity', code: 'AC' },
        { name: 'Privileges Required', code: 'PR' },
        { name: 'User Interaction', code: 'UI' },
        { name: 'Scope', code: 'S' },
        { name: 'Confidentiality Impact', code: 'C' },
        { name: 'Integrity Impact', code: 'I' },
        { name: 'Availability Impact', code: 'A' },
    ];

    // Mapping between the displayed option text and the corresponding CVSS code
    const mapping = {
        "Network": "N",
        "Adjacent Network": "A",
        "Local": "L",
        "Physical": "P",
        "High": "H",
        "Low": "L",
        "None": "N",
        "Required": "R",
        "Unchanged": "U",
        "Changed": "C",
    };

    // Function to get the CVSS value for a given section name
    function getCVSSValue(name) {
        // Find the span element that contains the section name
        const span = Array.from(document.querySelectorAll('span')).find(span => span.textContent.includes(name));
        if (span) {
            // Find the button within the same parent element that is pressed
            const button = span.parentElement.nextElementSibling.querySelector(`button[aria-pressed='true']`);
            return button ? button.textContent.trim() : null;
        }
        return null;
    }

    // Function to build the CVSS string based on selected values
    function buildCVSSString() {
        let cvssString = ''; // Initialize an empty string for the CVSS string
        // Iterate over each section to construct the CVSS string
        for (const section of sections) {
            const value = getCVSSValue(section.name); // Get the CVSS value for the section
            const code = mapping[value] || 'X'; // Get the corresponding code, default to 'X' if not found
            cvssString += `${section.code}:${code}/`; // Append the code to the CVSS string
        }
        displayCVSSString(cvssString.slice(0, -1)); // Remove the trailing slash and display the CVSS string
    }

    // Function to display the CVSS string on the page
    function displayCVSSString(cvssString) {
        const container = document.querySelector('div.col-md-3'); // Find the container to display the CVSS string
        if (!container) return; // Exit if the container doesn't exist
        let cvssElement = container.querySelector('.cvss-string'); // Find the existing CVSS string element
        if (!cvssElement) {
            // Create a new element if it doesn't exist
            cvssElement = document.createElement('span');
            cvssElement.className = 'cvss-string';
            cvssElement.style.color = 'yellow';
            container.appendChild(cvssElement); // Append the new element to the container
        }
        cvssElement.textContent = cvssString; // Set the text content to the CVSS string
    }

    // Initialize the script by attaching click listeners and building the initial CVSS string
    function init() {
        const buttons = document.querySelectorAll('.q-btn-group button'); // Find all the relevant buttons
        // Attach click listeners to each button to rebuild the CVSS string on click
        buttons.forEach(button => button.addEventListener('click', () => {
            setTimeout(buildCVSSString, 100); // Delay to allow the button state to update
        }));
        buildCVSSString(); // Build the initial CVSS string
    }

    // Use a MutationObserver to reinitialize the script if the page content changes
    const observer = new MutationObserver(init);
    observer.observe(document, { childList: true, subtree: true }); // Observe changes in the document
    init(); // Initial call to set up the script
})();

This should provide clarity on what each part of the script is doing, making it easier for others to understand and maintain.

Can you give upVote for this!!
